### ISSUE TITLE
Channel onUndeliveredElement doesn't work

### ISSUE THREAD

**Issue by artemkaxboy** [Kotlin/kotlinx.coroutines#3330]
I wanted to use channel with buffer but was disappointed to find out that it doesn't work as expected.

When I use a channel with **capacity = 1**, like this:

```kotlin
val capacity = 1

val channel = Channel<Int>(
    capacity = capacity,
    onBufferOverflow = BufferOverflow.DROP_OLDEST,
    onUndeliveredElement = { value -> println("Dropped value: $value") }
)

runBlocking {
    (1..3).forEach { value ->
        channel.send(value)
        println("Sent value: $value")
    }
}
```

It **works perfectly** as I want, output:

```
Sent value: 1
Dropped value: 1
Sent value: 2
Dropped value: 2
Sent value: 3
```

So I expected the same behavior with **greater capactity = 2**:

```kotlin
val capacity = 2

val channel = Channel<Int>(
    capacity = capacity,
    onBufferOverflow = BufferOverflow.DROP_OLDEST,
    onUndeliveredElement = { value -> println("Dropped value: $value") }
)

runBlocking {
    (1..3).forEach { value ->
        channel.send(value)
        println("Sent value: $value")
    }
}
```

But **onUndeliveredElement never called** in that case

```
Sent value: 1
Sent value: 2
Sent value: 3
```

BufferOverflow.DROP_LATEST doesn't work either.

**Comment by qwwdfsad** [comment]
This is indeed a bug, but it seems to be only fixable in 1.7.0.

The story is the following:

* `BufferOverflow` doesn't acknowledge `onUndeliveredElement` in any channels at all, except `CONFLATED`
* `Channel()` function has an internal optimization -- if capacity is `1` and strategy is `DROP_LATEST`, it picks `CONFLATED` as an underlying implementation because it's semantically the same.

So we have to fix all channels except `CONFLATED` which potentially can be considered as breaking change for most of the channels

**Comment by qwwdfsad** [comment]
While we're evaluating it, could you folks please describe your use case for this?

**Comment by artemkaxboy** [comment]
We have almost the same use case. We receive some data from the network and send items to a channel while the background worker processes them. When we have too many messages we'd like to drop the oldest. Planned to use callback for metrics and logging.

**Comment by qwwdfsad** [comment]
Fixed by #3621
